Update the README to use krane instead of kubernetes-deploy (#634)
Shopify/kubernetes-deploy: README.md

# krane [![Build status](https://badge.buildkite.com/61937e40a1fc69754d9d198be120543d6de310de2ba8d3cb0e.svg?branch=master)](https://buildkite.com/shopify/kubernetes-deploy) [![codecov](https://codecov.io/gh/Shopify/kubernetes-deploy/branch/master/graph/badge.svg)](https://codecov.io/gh/Shopify/kubernetes-deploy)

> This project used to be called `kubernetes-deploy`. Checkout our [migration guide](https://github.com/Shopify/kubernetes-deploy/blob/master/1.0-Upgrade.md) for more information including details about breaking changes.

Reviewer: ```suggestion
This project used to be called `kubernetes-deploy`. Check out our [migration guide](https://github.com/Shopify/kubernetes-deploy/blob/master/1.0-Upgrade.md) for more information including details about breaking changes.
```

​:running: [Running tasks at the beginning of a deploy](#running-tasks-at-the-beginning-of-a-deploy) using bare pods (example use case: Rails migrations)

> This repo also includes related tools for [running tasks](#krane run) and [restarting deployments](#krane restart).

Reviewer 2: We don't mention `krane render` in this section, even though you could argue it is going to be more important than it was in the `kubernetes-deploy` days, since otherwise there's no way to take advantage of the ERB templating

Reviewer: +1 definitely need to add a mention of `render` somewhere. Part of that should briefly mention how it can be used with deploy:
`krane render | krane deploy --stdin` sorta deal.

1. [Install kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/#install-kubectl-binary-via-curl) (requires v1.11.0 or higher) and make sure it is available in your $PATH
2. Set up your [kubeconfig file](https://kubernetes.io/docs/tasks/access-application-cluster/authenticate-across-clusters-kubeconfig/) for access to your cluster(s).
3. `gem install krane`

## Usage

`krane deploy <app's namespace> <kube context>`

*Environment variables:*

- `$KUBECONFIG`: points to one or multiple valid kubeconfig files that include the context you want to deploy to. File names are separated by colon for Linux and Mac, and semi-colon for Windows. If ommitted, will use the Kubernetes default of `~/.kube/config`.
- `$GOOGLE_APPLICATION_CREDENTIALS`: points to the credentials for an authenticated service account (required if your kubeconfig `user`'s auth provider is GCP)

*Options:*

Refer to `krane help` for the authoritative set of options.

> - `-f [PATHS]`: Accepts a list of directories and/or filenames to specify the set of directories/files that will be deployed.

Reviewer 2: We should include the longer `--filenames` here, as well, in accordance with the other entries

### Sharing a namespace

By default, krane will prune any resources in the target namespace which have the `kubectl.kubernetes.io/last-applied-configuration` annotation and are not a result of the current deployment process, on the assumption that there is a one-to-one relationship between application deployment and namespace, and that a deployment provisions all relevant resources in the namespace.

If you need to, you may specify `--no-prune` to disable all pruning behaviour, but this is not recommended.

If you need to share a namespace with resources which are managed by other tools or indeed other krane deployments, you can supply the `--selector` option, such that only resources with labels matching the selector are considered for pruning.

### Using templates and variables

Each app's templates are expected to be stored in a single directory. If this is not the case, you can create a directory containing symlinks to the templates. The recommended location for app's deploy directory is `{app root}/config/deploy/{env}`, but this is completely configurable.

> 

Reviewer 2: This is not strictly true anymore (though it is definitely a worth recommendation). Also, the header of this section mentions variables, but that's not what we talk about in the paragraph

> All templates must be YAML formatted.

Reviewer: Maybe it'd be worth it to mention "You may render ERB with `krane render` and then pipe that result to krane deploy --stdin`".

  that use the `RollingUpdate` strategy.
  - Percent (e.g. 90%): The deploy is successful when the number of new pods that are ready is equal to
  `spec.replicas` * Percent.
- `krane.shopify.io/predeployed`: Causes a Custom Resource to be deployed in the pre-deploy phase.
  - _Compatibility_: Custom Resource Definition
  - _Default_: `true`
  - `true`: The custom resource will be deployed in the pre-deploy phase.
  - All other values: The custom resource will be deployed in the main deployment phase.

### Running tasks at the beginning of a deploy

> To run a task in your cluster at the beginning of every deploy, simply include a `Pod` template in your deploy directory. `krane` will first deploy any `ConfigMap` and `PersistentVolumeClaim` resources in your template set, followed by any such pods. If the command run by one of these pods fails (i.e. exits with a non-zero status), the overall deploy will fail at this step (no other resources will be deployed).

Reviewer 2: 'template set' is more of an internal detail of `krane` as opposed to a top level concern. From a user's point of view, they are simply passing in a list of filenames/dirs. It might be more accurate to say `resources from files provided to krane`.

:man_shrugging: I'm not too hung up on it, though

**Note: If you're a Shopify employee using our cloud platform, this setup has already been done for you. Please consult the CloudPlatform User Guide for usage instructions.**

Since their data is only base64 encoded, Kubernetes secrets should not be committed to your repository. Instead, `krane` supports generating secrets from an encrypted [ejson](https://github.com/Shopify/ejson) file in your template directory. Here's how to use this feature:

1. Install the ejson gem: `gem install ejson`
2. Generate a new keypair: `ejson keygen` (prints the keypair to stdout)
3. Create a Kubernetes secret in your target namespace with the new keypair: `kubectl create secret generic ejson-keys --from-literal=YOUR_PUBLIC_KEY=YOUR_PRIVATE_KEY --namespace=TARGET_NAMESPACE`
>Warning: Do *not* use `apply` to create the `ejson-keys` secret. krane will fail if `ejson-keys` is prunable. This safeguard is to protect against the accidental deletion of your private keys.
4. (optional but highly recommended) Back up the keypair somewhere secure, such as a password manager, for disaster recovery purposes.

> 5. In your template directory (alongside your Kubernetes templates), create `secrets.ejson` with the format shown below. The `_type` key should have the value “kubernetes.io/tls” for TLS secrets and “Opaque” for all others. The `data` key must be a json object, but its keys and values can be whatever you need.

Reviewer 2: Should we note that EJSON secrets _cannot_ be passed via STDIN (at least I'm pretty sure they can't be)

Reviewer: IMO We should be sure and be explicit about it.

    message: "resource is failed"
```

- `observedGeneration == metadata.generation`, so krane will check this resource's success and failure conditions.
- Since `$.status.conditions[?(@.type == "Ready")].status == "False"`, the resource is not considered successful yet.
- `$.status.conditions[?(@.type == "Failed")].status == "True"` means that a failure condition has been fulfilled and the resource is considered failed.
- Since `error_msg_path` is specified, krane will log the contents of `$.status.conditions[?(@.type == "Failed")].message`, which in this case is: `resource is failed`.

### Deploy walkthrough

Let's walk through what happens when you run the `deploy` task with [this directory of templates](https://github.com/Shopify/kubernetes-deploy/tree/master/test/fixtures/hello-cloud). You can see this for yourself by running the following command:

```bash

> krane render -f test/fixtures/hello-cloud --current-sha 1| krane deploy my-namespace my-k8s-cluster  --stdin

Reviewer 2: This reference to `krane render` comes out of left field and will probably be confusing to people reading through this for the first time. Minimally, I think we should provide a link to the section on `krane render` and a note/comment that, since these fixtures use ERB templates, we must first pass them through our renderer

**On timeouts**: It's important to notice that a single resource timeout or a global deploy timeout doesn't necessarily mean that the operation failed. Since Kubernetes updates are asynchronous, maybe something was just too slow to return in the configured time; in those cases, usually running the deploy again might work (that should be a no-op for most - if not all - resources).

# krane global deploy

Ship non-namespaced resources to a cluster

krane global-deploy (accessible through the Ruby API as Krane::GlobalDeployTask) can deploy global (non-namespaced) resources such as PersistentVolume, Namespace, and CustomResourceDefinition.
Its interface is very similar to krane deploy.

## Usage

`krane global-deploy <kube context>`

```bash
$ cat my-template.yml
    apiVersion: storage.k8s.io/v1
    kind: StorageClass
    metadata:
      name: testing-storage-class
      labels:
        app: krane
    provisioner: kubernetes.io/no-provisioner

$ krane global-deploy my-k8s-context -f my-template.yml --selector app=krane
```

*Options:*

Refer to `krane global-deploy help` for the authoritative set of options.

> - `--filenames / -if [PATHS]`: Accepts a  list of directories and/or filenames to specify the set of directories/files that will be deployed

Reviewer: ```suggestion
- `--filenames / -if [PATHS]`: Accepts a list of directories and/or filenames to specify the set of directories/files that will be deployed
```

* The `Pod` specification in that template has a container named `task-runner`.

Based on this specification `krane run` will create a new pod with the entrypoint of the `task-runner ` container overridden with the supplied arguments.

## Usage

`krane run <kube namespace> <kube context> --arguments=<arguments> --command=<command> --template=<template name>`

*Options:*

* `--template=TEMPLATE`:  Specifies the name of the PodTemplate to use (default is `task-runner-template` if this option is not set).
* `--env-vars=ENV_VARS`: Accepts a list of environment variables to be added to the pod template. For example, `--env-vars="ENV=VAL ENV2=VAL2"` will make `ENV` and `ENV2` available to the container.
* `--command=`: Override the default command in the container image.
* `--no-verify-result`: Skip verification of pod success
* `--global-timeout=seconds`: Raise a timeout error if the pod runs for longer than the specified number of seconds
* `--arguments:`: Override the default arguments for the command with a space-separated list of arguments

# krane render

> `krane render` is a tool for rendering ERB templates to raw Kubernetes YAML. It's useful for outputting YAML that can be passed to other tools, for validation or introspection purposes. It can also be used.

Reviewer: Looks like you didn't finish a thought at the end of this paragraph

* `--no-verify-result`: Skip verification of pod success
* `--global-timeout=duration`: Raise a timeout error if the pod runs for longer than the specified duration
* `--arguments:`: Override the default arguments for the command with a space-separated list of arguments

# krane render

`krane render` is a tool for rendering ERB templates to raw Kubernetes YAML. It's useful for outputting YAML that can be passed to other tools, for validation or introspection purposes.

## Prerequisites

 * `krane render` does __not__ require a running cluster or an active kubernetes context, which is nice if you want to run it in a CI environment, potentially alongside something like https://github.com/garethr/kubeval to make sure your configuration is sound.

## Usage

To render all templates in your template dir, run:

```
krane render -f ./path/to/template/dir
```

To render some templates in a template dir, run krane render with the names of the templates to render:

```
krane render -f ./path/to/template/dir/this-template.yaml.erb
```

To render a template in a template dir and output it to a file, run krane render with the name of the template and redirect the output to a file:

```

> krane render --f./path/to/template/dir/template.yaml.erb > template.yaml

Reviewer: ```suggestion
krane render -f ./path/to/template/dir/template.yaml.erb > template.yaml
```
